Source/NSPathControl.m

```diff
@@ -0,0 +1,565 @@
+
+#import "AppKit/NSPathControl.h"
+#import "AppKit/NSPathCell.h"
+#import "AppKit/NSGraphics.h"
+#import "AppKit/NSDragging.h"
+#import "AppKit/NSPasteboard.h"
+#import "AppKit/NSMenu.h"
+#import "AppKit/NSOpenPanel.h"
+#import "AppKit/NSPathComponentCell.h"
+#import "AppKit/NSPathControlItem.h"
+#import "AppKit/NSEvent.h"
+#import "AppKit/NSTrackingArea.h"
+
+static NSNotificationCenter *nc = nil;
+static Class pathCellClass;
+
+@interface NSPathCell (PathControlPrivate)
+- (void) _setClickedPathComponentCell: (NSPathComponentCell *)c;
+@end
+
+@interface NSPathComponentCell (PathControlPrivate)
+- (NSPathControlItem *) _pathControlItem;
+@end
+
+@implementation NSPathControl
+
++ (void) initialize
+{
+  if (self == [NSPathControl class])
+    {
+      [self setVersion: 1.0];
+      [self setCellClass: [NSPathCell class]];
+      nc = [NSNotificationCenter defaultCenter];
+    }
+}
+
+
++ (Class) cellClass
+{
+  return pathCellClass;
+}
+
++ (void) setCellClass: (Class)classId
+{
+  pathCellClass = classId;
+}
+
+- (void) resetCursorRects
+{
+  [[self superview] removeTrackingRect: _trackingTag];
+  _trackingTag = [[self superview] addTrackingRect: [self frame]
+                                             owner: self
+                                          userData: nil
+                                      assumeInside: YES];
+}
+
+- (instancetype) init
+{
+  self = [super init];
+  if (self != nil)
+    {
+      [self setPathStyle: NSPathStyleStandard];
+      [self setURL: nil];
+      [self setDelegate: nil];
+      [self setAllowedTypes: [NSArray arrayWithObject: NSFilenamesPboardType]];
+      [self resetCursorRects];
+    }
+  return self;
+}
+
+- (void) dealloc
+{
+  [[self superview] removeTrackingRect: _trackingTag];
+  [super dealloc];
+}
+
+- (void) mouseEntered: (NSEvent *)event
+{
+  [_cell mouseEntered: event
+            withFrame: [self frame]
+               inView: self];
+}
+
+- (void) mouseExited: (NSEvent *)event
+{
+  [_cell mouseExited: event
+           withFrame: [self frame]
+              inView: self];
+}
+
+- (void) setPathStyle: (NSPathStyle)style
+{
+  [_cell setPathStyle: style];
+  [self setNeedsDisplay];
+}
+
+- (NSPathStyle) pathStyle
+{
+  return [_cell pathStyle];
+}
+
+- (NSPathComponentCell *) clickedPathComponentCell
+{
+  return [_cell clickedPathComponentCell];
+}
+
+- (NSArray *) pathComponentCells
+{
+  return [_cell pathComponentCells];
+}
+
+- (void) setPathComponentCells: (NSArray *)cells
+{
+  [_cell setPathComponentCells: cells];
+  [self setNeedsDisplay];
+}
+
+- (SEL) doubleAction;
+{
+  return [_cell doubleAction];
+}
+
+- (void) setDoubleAction: (SEL)doubleAction
+{
+  [_cell setDoubleAction: doubleAction];
+}
+
+- (void) _createPathItems
+{
+  NSArray *a = [_cell pathComponentCells];
+  NSEnumerator *en = [a objectEnumerator];
+  NSPathComponentCell *c = nil;
+  NSMutableArray *items = [NSMutableArray arrayWithCapacity: [a count]];
+  
+  while((c = [en nextObject]) != nil)
```

Comment:
Again a space after the "while" is missing.

Reply:
Sorry, but the space is still missing. Why do you resolve the conversation without resolving the issue?

Reply:
I dont know how i missed it.  Fixed.